V1.3 sort command

ZhiWei1010 opened the pull request:
This commit adds a new sort feature to PatientSync.
* Add new SortCommand class
* Add new SortCommandParser class
* Add new NameComparator class to support the sorting by patient name
* Add new PreferredNameComparator class to support the sorting by patient's preferred name
* Update AddressBookParser to recognise `sort` command
* Add new SortCommandTest class
* Add new SortCommandParserTest class
* Add new test case in AddressBookParser to include test for sort command
* Update User Guide to include usage of Sort Command
* Add new sort result screenshot
* Update Developer Guide to include section on Sort Command
* Add new Sequence Diagram for Sort Command (SortSequenceDiagram.puml)
* Add new Activity Diagram for Sort Command (SortActivityDiagram.puml)

Fixes #159

ZhiWei1010 committed:
Logic: Add new SortCommand

Adds new SortCommand class to handle sorting of patient list.

Adds new SortCommand parser to handle parsing of sort command and user
input arguments.

Adds new case in AddressBookParser#parseCommand(String userInput) to
detect sortCommand.

ZhiWei1010 committed:
Model: Updates Model to support sort command

Add new NameComparator class to compare patients by name.

Add new PreferredNameComparator class to compare patients by preferred
name.

Add new Model#getFullPatientList() method to get unmodifiable view of
the full patient list.

Add new Model#updatePatientList(List<Patient>) method to update the
patient list with an updated patient list.

Add new ModelManager#getFullPatientList() to get the full patient list.

Add new ModelManager#updatePatientList(List<Patient>) to updated the
patient list with an updated patient list.

Update the ModelStub in AddCommandTest to override new methods added in
Model Interface.

ZhiWei1010 committed:
Logic: Update SortCommand class

Add new parameter `sortAttribute` for SortCommand constructor.

Update SortCommandParser to pass sortAttribute to SortCommand.

ZhiWei1010 committed:
Logic: Update SortCommand#equals() and SortCommand#toString()

Update SortCommand#equals() to check for same sort attribute too.

Update SortCommand#toString to include sort attribute.

ZhiWei1010 committed:
Logic: Add new assert statements in SortCommand

Add new assert statements in SortCommand to ensure that sort attribute
matches with comparator.

ZhiWei1010 committed:
Test: Add new SortCommandTest and SortCommandParserTest

Add new SortCommandTest class to test for sort command.

Add new SortCommandParserTest class to test for sort command parser.

ZhiWei1010 committed:
Test: Update SortCommandTest and SortCommandParser to fix failing test

ZhiWei1010 committed:
Logic: Update SortCommand Bug

Change all instances of PREFIX_PREFERRED_NAME to
PREFERRED_NAME_COMPARATOR in SortCommand.

Update example of MESSAGE_USAGE to correct example.

ZhiWei1010 committed:
Test: Update AddressBookParserTest and SortCommandTest

Add test for sort command in AddressBookParserTest.

Add tests in SortCommandTest to test for mismatch sort attributes
and comparators.

ZhiWei1010 committed:
Logic: Update SortCommandParser to throw new AssertionError at
unreachable code

ZhiWei1010 committed:
Test: Add new test case in SortCommandTest#toStringTest()

ZhiWei1010 committed:
Docs: Update UserGuide to include SortCommand

ZhiWei1010 committed:
Docs: Update DeveloperGuide.md to include SortCommand

Add new section in DeveloperGuide to include specifications of
SortCommand.

Add new sequence diagram for SortCommand.

Add new activity diagram for SortCommand.

ZhiWei1010 committed:
Docs: Add Screenshot for sort result in UserGuide.md

Add a new screenshot in the image folder for sort result.

Update the User Guide to display the sort result screenshot.

choowengyan commented on docs/DeveloperGuide.md:
```diff
@@ -781,6 +781,89 @@ The following UML sequence diagram illustrates how the FindTags operations works
+
+Step 1: The user accesses the PatientSync application.
+
+Step 2: The user executes `sort p` to sort patient list by patient's preferred name.
```
Would it be better to also include example of sorting patient list by patient's name here as well?

ZhiWei1010 replied:
Sorting by patient name follows a similar flow of events, thus I feel that showing one example should be sufficient!

choowengyan commented on docs/images/sortResult.png:
Will the image be better if it shows the input command line together with the successful message?

ZhiWei1010 replied:
I think it is not necessary as the command is not meant to show after entering the command! It is also consistent with the other result images for the other commands in the User Guide!

choowengyan reviewed (changes requested):
Overall, good job on the sort command! May require slight modification on the documentation for clarity purpose.

SherisseTJW commented on src/main/java/seedu/address/logic/commands/SortCommand.java:
```diff
@@ -0,0 +1,101 @@
+
+    public static final String COMMAND_WORD = "sort";
+
+    public static final String MESSAGE_SORT_SUCCESS = "Successfully sorted all patients by ";
```
Maybe follow the way the other commands do it, which is to use `%1$s` and `String.format()`

ZhiWei1010 replied:
Good catch! Will proceed to change it!

SherisseTJW commented on src/main/java/seedu/address/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,79 @@
+        logger.log(Level.INFO, "trimmed argument: " + trimmedArgs);
+
+        if (trimmedArgs.length() > 1) {
+            logger.log(Level.WARNING, "Invalid argument length (> 1) received.");
```
Probably rename the logger to say `Invalid Attribute length` instead? since the docs all refer to this as the Attribute, not the argument. Either that or standardise to Argument throughout

ZhiWei1010 replied:
Good catch! Will proceed to update it to invalid attribute length!

SherisseTJW reviewed (commented):
generally logic is fine, but maybe can standardise a bit more to the other commands / docs that we already have

NeoHW commented on docs/DeveloperGuide.md:
```diff
@@ -781,6 +781,89 @@ The following UML sequence diagram illustrates how the FindTags operations works
+
+#### Specifications
+
+* `SortCommand` takes in one zero attributes to sort the patient list.
```
one zero?

NeoHW replied:
do u mean one or zero

ZhiWei1010 replied:
Good catch! the correct one is one or zero!

NeoHW commented on docs/DeveloperGuide.md:
```diff
@@ -781,6 +781,89 @@ The following UML sequence diagram illustrates how the FindTags operations works
+
+**Aspect: Method to update the model after sorting**
+
+* **Alternative 1 (current choice)**: Create new `updatePatientList` method in Model
+  * Pros: Updates the patient list with little code.
+  * Cons: Need to ensure that the sorted patient list does not add or deletes patients accidentally, else
+there will be extra or missing patients.
+    <br><br>
+* **Alternative 2**: Sort the patients, deletes all the patients in the patient list, then add the patients back in the
+sorted order.
+  * Pros: Easy and simple.
+  * Cons: Slow and inefficient.
```
was wondering if this aspect is needed since pre-existing methods already uses current choice instead of deleting and adding back patients

ZhiWei1010 replied:
Pre-existing methods uses updateFilteredPatientList if Im not wrong and there is no updatePatientList method in Model and ModelManager prior to SortCommand's implementation!

NeoHW commented on docs/UserGuide.md:
```diff
@@ -324,6 +325,35 @@ Examples:
+
+* Sorts all the added patients in the patient list by specified `ATTRIBUTE`.
+* Supported `ATTRIBUTE`s include `n` for patient name and `p` for patient's preferred name.
+* There can only be 1 or 0 `ATTRIBUTE`.
```
DG uses one and zero, 
could consider standardising the UG and DG to one of the options

ZhiWei1010 replied:
Good catch! will proceed to standardise to one or zero!

NeoHW commented on docs/UserGuide.md:
```diff
@@ -324,6 +325,35 @@ Examples:
+* Sorts all the added patients in the patient list by specified `ATTRIBUTE`.
+* Supported `ATTRIBUTE`s include `n` for patient name and `p` for patient's preferred name.
+* There can only be 1 or 0 `ATTRIBUTE`.
+* If no `ATTRIBUTE` is specified, the `sort` command will wort the patient list by patient's name by default.
```
~wort~ sort

ZhiWei1010 replied:
Good Catch! will proceed to fix the spelling error!

NeoHW commented on docs/UserGuide.md:
```diff
@@ -324,6 +325,35 @@ Examples:
+* Sorts all the added patients in the patient list by specified `ATTRIBUTE`.
+* Supported `ATTRIBUTE`s include `n` for patient name and `p` for patient's preferred name.
+* There can only be 1 or 0 `ATTRIBUTE`.
+* If no `ATTRIBUTE` is specified, the `sort` command will wort the patient list by patient's name by default.
```
could consider rephrasing it to " if no attribute is specified, default sorting order would be by patient's name "
to avoid redundancy in sentence

ZhiWei1010 replied:
Good suggestion! Will proceed to update the sentence!

ZhiWei1010 committed:
Logic: Minor update in naming convention in SortCommandParser

Update naming convention of trimmedArgs to trimedSortAttribute in
SortCommandParser.

Update MESSAGE_SORT_SUCCESS format in SortCommand.

ZhiWei1010 committed:
Test: Update test case in SortCommandTest to follow new format
for success message

NeoHW commented on src/main/java/seedu/address/logic/commands/SortCommand.java:
```diff
@@ -0,0 +1,101 @@
+        logger.log(Level.INFO, "Successfully sorted the patient list by " + this.sortAttribute);
+
+        return new CommandResult(MESSAGE_SORT_SUCCESS + sortAttribute);
+    }
```
Considering that the Comparator and sortAttribute are inherently linked, is there a necessity to pass both and then validate through assertions by comparing them? The input is determined by the parser logic. Instead, it might be simpler to pass in just one, and then create the corresponding Comparator or sortAttribute accordingly. This approach could potentially reduce the likelihood of errors occurring.

ZhiWei1010 replied:
Good suggestion! That was my initial design for the code where I would get the sortAttribute based on checks on the Comparator, however I felt that since I am parsing the sort attribute in the sortCommandParser, I could get the sort attribute there, so I decided to change it.

But now that I revisit this code, I think it would be better to create the corresponding sortAttribute to reduce the likelihood of errors occurring! Will proceed to change it!

NeoHW commented on src/main/java/seedu/address/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,79 @@
+        case "n":
+            sortAttribute = "name";
+            logger.log(Level.INFO, "Sorting Attribute: Patient " + sortAttribute);
+            return new SortCommand(NAME_COMPARATOR, sortAttribute);
```
could use fallthrough here

ZhiWei1010 replied:
The checkstyle fails when fallthrough is used, thus the reason for adding additional return statement in `case ""`!

NeoHW commented on src/main/java/seedu/address/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,79 @@
+        case "p":
+            sortAttribute = "preferred name";
+            logger.log(Level.INFO, "Sorting Attribute: Patient's " + sortAttribute);
+            return new SortCommand(PREFERRED_NAME_COMPARATOR, sortAttribute);
```
refer to comment in SortCommand for suggestion

ZhiWei1010 replied:
refer to reply in SortCommand!

NeoHW reviewed (changes requested):
Some changes to be made, with some design considerations

ZhiWei1010 committed:
Docs: Update to DeveloperGuide and UserGuide

Add new user story in DeveloperGuide.

Update UserGuide SortCommand to be consistent with DeveloperGuide.

ZhiWei1010 committed:
Logic: Update design of SortCommand

SortCommand now checks the comparator to get the respective
sortAttribute instead of accepting sortAttribute as a parameter.

SortCommandParser now only pass the Comparator to create SortCommand.

ZhiWei1010 committed:
Test: Update SortCommandTest, SortCommandParserTest and
AddressBookParserTest to accommodate for new design of SortCommand
and SortCommandParser

NeoHW reviewed (approved):
lgtm

choowengyan reviewed (approved):
LGTM!
